text/0041-fixed-point.md

```diff
@@ -0,0 +1,104 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [amaranth-lang/rfcs#41](https://github.com/amaranth-lang/rfcs/pull/41)
+- Amaranth Issue: [amaranth-lang/amaranth#0000](https://github.com/amaranth-lang/amaranth/issues/0000)
+
+# Fixed point types
+
+## Summary
+[summary]: #summary
+
+Add fixed point types to Amaranth.
+
+## Motivation
+[motivation]: #motivation
+
+Fractional values in hardware are usually represented as some form of fixed point value.
+Without a first class fixed point type, the user has to manually manage how the value needs to be shifted to be represented as an integer and keep track of how that interacts with arithmetic operations.
+
+A fixed point type would encode and keep track of the precision through arithmetic operations, as well as provide standard operations for converting values to and from fixed point representations with correct rounding.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+TODO
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes a library addition `amaranth.lib.fixedpoint` with the following contents:
+
+`FixedPoint` is a `ShapeCastable` subclass.
+The following operations are defined on it:
+
+- `FixedPoint(f_width, /, *, signed)`: Create a `FixedPoint` with `f_width` fractional bits.
+- `FixedPoint(i_width, f_width, /, *, signed)`: Create a `FixedPoint` with `i_width` integer bits and `f_width` fractional bits.
```

Comment:
These two lines seem mutually incompatible.

Reply:
The idea is that the underlying implementation would be `FixedPoint(i_or_f_width, f_width = None, /, *, signed)`, but I found it clearer to express it like that. Consider how you can do `range(stop)` and `range(start, stop)`.

Reply:
Makes sense. I wasn't sure if it was that or a typo. Does the first constructor add a sign bit when `signed` is true, and otherwise create a number whose range spans from 0 to some value below 1?

Reply:
Yes, although the exact semantics wrt. the sign bit depends on which Q notation we settle on.

Reply:
The dominant notation in the industry, at least in the audio ASIC world, is to include the sign bit in the number of integer bits. For example, -1 to 1 is represented as a Q1.23. That would be my vote.

Reply:
@samimia-swks would you happen to have some references we can cite to show that this is the dominant Q notation in the audio ASIC world? It will be useful evidence that may be worth appending to the RFC document itself.